### PR TITLE
server: Remove angle-bracket envoy includes, and make check_format prevent them from reappearing

### PR DESCRIPTION
**Pull request by jmarantz**
Signed-off-by: Joshua Marantz <jmarantz@google.com>

*Description*:
Angle-brackets are for system includes, generally, but
```
#include <envoy/server/config_tracker.h>
```
crept into our system.  That fails importing into Google which is why I noticed it.

*Risk Level*: Low

*Testing*: //test/...

*Docs Changes*: N/A
*Release Notes*: N/A

### CODE REVIEW

File: tools/check_format.py [hunk 1]
```diff
@@ -86,15 +86,19 @@ def isBuildFile(file_path):
 
 
 def checkFileContents(file_path):
-  message = "%s has over-enthusiastic spaces:" % file_path
-  findSubstringAndPrintError('.  ', file_path, message)
+  findSubstringAndPrintError('.  ', file_path, "over-enthusiastic spaces")
+  findSubstringAndPrintError('#include <envoy', file_path,
```
Comment (mattklein123) [inline tools/check_format.py]: Doesn't this basically apply to all project includes? Not just "envoy" ones?

Comment (htuch) [reply tools/check_format.py]: Yeah, we want `<test` and `<common` as well I think.

Comment (mattklein123) [reply tools/check_format.py]: And extensions now. Can we potentially somehow share the list with the header order fixer?

Comment (mattklein123) [reply tools/check_format.py]: (common py include potentially?)

Comment (jmarantz) [reply tools/check_format.py]: Good call.  in header_order.py there's a similar list, expressed as:
```
      file_header_filter(),
      regex_filter('<.*\.h>'),
      regex_filter('<.*>'),
      regex_filter('"envoy/.*"'),
      regex_filter('"common/.*"'),
      regex_filter('"source/common/.*"'),
      regex_filter('"exe/.*"'),
      regex_filter('"server/.*"'),
      regex_filter('"extensions/.*"'),
      regex_filter('"test/.*"'),
```

The 7 of these could be generated from a shared array of strings.  Would it be cool to effectively change "/source/common/.*" above to just "/source/.*"?  Specifically, if someone loads from source/server what should happen?   It seems like that maybe the intent is that it shouldn't pass lint, and it should not be auto-fixable either.  Or we could -- at the tooling level -- just ignore that.

Comment (mattklein123) [reply tools/check_format.py]: Yeah I think that would be fine.

Comment (jmarantz) [reply tools/check_format.py]: done

File: tools/common.py [hunk 2]
```diff
@@ -0,0 +1,10 @@
+def include_dir_order():
```
Comment (htuch) [inline tools/common.py]: Nit: This is the third example of function name capitalization style in this set of related files.

File: tools/check_format.py [hunk 3]
```diff
@@ -86,15 +87,22 @@ def isBuildFile(file_path):
...
+    findSubstringAndPrintError(prefix, file_path, "envoy includes should not have angle brackets")
 
 
 def fixFileContents(file_path):
   for line in fileinput.input(file_path, inplace=True):
     # Strip double space after '.'  This may prove overenthusiastic and need to
     # be restricted to comments and metadata files but works for now.
-    sys.stdout.write("%s" % (line.replace('.  ', '. ')))
+    line = line.replace('.  ', '. ')
+    for subdir in common.include_dir_order():
```
Comment (htuch) [inline tools/check_format.py]: This seems like a lot of work to do for every line. We should guard it with a check that only lines with `#include <` prefix are under it.

File: tools/common.py [hunk 4]
```diff
@@ -0,0 +1,10 @@
...
+  return (
+      "envoy/",
```
Comment (htuch) [inline tools/common.py]: Can this just be a constant set? That way we have fast hash lookup.

File: tools/check_format.py (hunk 3 above)
Comment (jmarantz) [reply tools/check_format.py]: oh that's a good idea, and in fact there's a lot of performance improvement we can/should do.  Will update.

File: tools/common.py (hunk 4 above)
Comment (jmarantz) [reply tools/common.py]: Well the order is significant in one usage.  Making a dict makes sense though, not here but in the usage in check_format.py.

File: tools/common.py [hunk 5]
```diff
@@ -0,0 +1,10 @@
+def includeDirOrder():
+  return (
+      "envoy",
+      "common",
+      "source",
```
Comment (mattklein123) [inline tools/common.py]: Should we just block including using "source/"? @htuch do you recall how this ended up here? I don't remember now.

Comment (jmarantz) [reply tools/common.py]: recursive grep tells me there are exactly two files in the repo that include "source/..":

./source/common/ratelimit/ratelimit_impl.h:#include "source/common/ratelimit/ratelimit.pb.h"
./test/integration/ratelimit_integration_test.cc:#include "source/common/ratelimit/ratelimit.pb.h"

Comment (jmarantz) [reply tools/common.py]: Also I want to suggest not having a policy-change like this be a new responsibility of this PR...happy to hack that as a follow up if you like :)

Comment (mattklein123) [reply tools/common.py]: Sure makes sense. I was just curious.

**Comment by htuch** [comment]
Was chatting offline with @jmarantz and I think the plan is to move this functionality into `header_order.py`.

**Comment by htuch** [comment]
Actually, based on more offline convos, `header_order.py` doesn't really help, so let's continue as is.

**Comment by jmarantz** [comment]
Any more comments?

**Comment by jmarantz** [comment]
Just to provide more context for the logs, the reason to keep this as is, as opposed to moving it to header_order.py, is that we need to check/fix the syntax of the includes prior to doing semantic analysis.  It's not too bad to do syntactic include-syntax analysis in check_format.py, where the structure, as of this PR is:
```
foreach file:
  foreach line:
    foreach rule:
      apply rule
```
It's not that wasteful either to check for ```"#include <"``` in check_format.py, even though we are also checking for ```"#include "``` in header_order.py. 

To correctly put all this functionality into header_order.py we'd need to add a new syntactic pass at the beginning of header_order.py, which would really duplicate the structure above.

File: tools/check_format.py [hunk 6]
```diff
@@ -85,16 +86,46 @@ def isBuildFile(file_path):
...
 
 
-def checkFileContents(file_path):
-  message = "%s has over-enthusiastic spaces:" % file_path
-  findSubstringAndPrintError('.  ', file_path, message)
+def headerSubdirDict():
+  subdir_dict = {}
+  for subdir in common.includeDirOrder():
+    subdir_dict[subdir] = True
+  return subdir_dict
```
Comment (htuch) [inline tools/check_format.py]: I would just make this `headerSubdirSet` and then this entire function becomes `return set(common.includeDirOrder())`.

File: tools/check_format.py [hunk 7]
```diff
@@ -85,16 +86,46 @@ def isBuildFile(file_path):
...
+    return False
+  subdir = path[0:slash]
+  return SUBDIR_DICT.has_key(subdir)
 
+def printLineError(path, line_number, message):
+  printError("%s:%d: %s" % (path, line_number, message))
+
+def checkFileContents(file_path):
+  line_number = 0
+  for line in fileinput.input(file_path):
```
Comment (htuch) [inline tools/check_format.py]: `for line_number, line in enumerate(fileinput.input(file_path)):`

File: tools/check_format.py (hunk 6 above)
Comment (jmarantz) [reply tools/check_format.py]: done -- actually that became simple enough I just did that in the const initializer.

File: tools/check_format.py (hunk 7 above)
Comment (jmarantz) [reply tools/check_format.py]: done...annoyingly they are zero-based.

**Pull request merged**